### PR TITLE
Fade volume

### PR DESCRIPTION
**Pull request by ninovanhooff**

### CODE REVIEW

File: src/playdate/bindings/sound.nim [hunk 1]
```diff
@@ -38,15 +38,15 @@ type PlaydateSoundFileplayer {.importc: "const struct playdate_sound_fileplayer"
...
+        player: FilePlayerPtr; callback: PDSndCallbackProcRaw, userdata: pointer = nil) {.cdecl, raises: [].}
     # setLoopCallback* {.importc: "setLoopCallback".}: proc (player: ptr FilePlayer;
     #     callback: SndCallbackProc) {.cdecl.}
     getOffset {.importc: "getOffset".}: proc (player: FilePlayerPtr): cfloat {.cdecl, raises: [].}
     # getRate* {.importc: "getRate".}: proc (player: ptr FilePlayer): cfloat {.cdecl.}
     # setStopOnUnderrun* {.importc: "setStopOnUnderrun".}: proc (
     #     player: ptr FilePlayer; flag: cint) {.cdecl.}
-    # fadeVolume* {.importc: "fadeVolume".}: proc (player: ptr FilePlayer; left: cfloat;
-    #     right: cfloat; len: int32T; finishCallback: SndCallbackProc) {.cdecl.}
+    fadeVolume* {.importc: "fadeVolume".}: proc (player: FilePlayerPtr; left: cfloat;
```
Comment (ninovanhooff) [inline src/playdate/bindings/sound.nim:48]: For parity's sake, we might want to implement this for PlaydateSoundSampleplayer too

Comment (Nycto) [reply src/playdate/bindings/sound.nim:48]: We can do this later, imo.

File: src/playdate/sound.nim [hunk 2]
```diff
@@ -139,6 +140,11 @@ proc privateFilePlayerFinishCallback(soundSource: SoundSourcePtr, userdata: poin
     if filePlayer.finishCallback != nil:
         filePlayer.finishCallback(filePlayer)
 
+proc privateFilePlayerFadeVolumeCallback(soundSource: SoundSourcePtr, userdata: pointer) {.cdecl, raises: [].} =
```
Comment (Nycto) [inline src/playdate/sound.nim:143]: Should we be doing something with SoundSourcPtr? Passing that into callback, for example?

File: src/playdate/sound.nim [hunk 3]
```diff
@@ -148,6 +154,15 @@ proc setFinishCallback*(this: FilePlayer, callback: PDFilePlayerCallbackFunction
...
         playdate.sound.fileplayer.setFinishCallback(this.resource, privateFilePlayerFinishCallback, cast[pointer](this))
 
+proc fadeVolume*(this: FilePlayer, left, right: float32, len: int32, callback: PDFilePlayerCallbackFunction) =
+    privateAccess(PlaydateSound)
+    privateAccess(PlaydateSoundFileplayer)
+    this.fadeVolumeCallback = callback
+    if callback == nil:
+        playdate.sound.fileplayer.fadeVolume(this.resource, left.cfloat, right.cfloat, len.cint , nil, nil)
+    else:
+        playdate.sound.fileplayer.fadeVolume(this.resource, left.cfloat, right.cfloat, len.cint , privateFilePlayerFadeVolumeCallback, cast[pointer](this))
```
Comment (Nycto) [inline src/playdate/sound.nim:164]: The api docs don't mention that last parameter:

https://sdk.play.date/2.6.0/Inside%20Playdate%20with%20C.html#f-sound.fileplayer.fadeVolume

Have you tried this?

Comment (ninovanhooff) [reply src/playdate/sound.nim:164]: Yes, in Wheelsprung I use both with and without the last param.

Reported here:
https://devforum.play.date/t/c-docs-fadevolume-is-missing-userdata/20245

File: src/playdate/sound.nim (hunk 2 above)
Comment (ninovanhooff) [reply src/playdate/sound.nim:143]: I'd say this is an internal type that we don't want to expose to game code?

```
type SoundSourcePtr = ptr SoundSourceRaw
```

**Review by Nycto — APPROVED** [review]

**Pull request merged**